### ISSUE TITLE
Guidance on the lifetime of the zoom client

### ISSUE THREAD

**Issue by ashilon** [Jericho/ZoomNet#35]
Hi,
Adding some guidelines and best practices about lifetime management of the Zoom client would be very helpful, especially in asp.net core web apps where we need to determine the lifetime of zoom object when adding it to the di services in asp.net core.
Thank you very much and your work on this library is very much appreciated.

**Comment by Jericho** [comment]
By default, the Zoom client creates an instance of `HttpClient` in order to send requests to the Zoom API. This means that the lifetime guidance for ZoomClient is the same as the lifetime guidance for HttpClient. Microsoft's documentation states the following:

> HttpClient is intended to be instantiated once and re-used throughout the life of an application. Instantiating an HttpClient class for every request will exhaust the number of sockets available under heavy loads. This will result in SocketException errors.

Please note that ZoomClient has a constructor that allows you to provide you own HttpClient and therefore you could instantiate a new ZoomClient with every request as long as you provide your own HttpClient and you properly manage the lifetime of this HttpClient.

One final consideration: if you use OAuth you should definitely re-use the same instance of the ZoomClient. The reason for this recommendation is that ZoomClient makes a request to the Zoom API to obtain an authorization token when it gets instantiated and renews this token when it expires every 60 minutes. This means that ZoomClient issues one http request to the Zoom API per hour if you re-use the same instance but if you decide to instantiate a new ZoomClient every time you need to make a call to the Zoom API, the ZoomClient will make a call to get the authorization token every single time.

In summary:

- It's best to instantiate the ZoomClient once and re-use it.
- You can instantiate a new ZoomClient with every request if you provide your own HttpClient and you use JWT but **definitely don't do this if you use OAuth**.

Let me know if this helps.

**Comment by ashilon** [comment]
This info helps a lot. You have described the two problematic issues, OAuth access token reuse and HttpCient usage, so these topics are very clear now. Thanks a bunch.

**Comment by Jericho** [comment]
> Since the token is specific to each end user, does ZoomClient store it in asp.core's session state?

The ZoomClient do not manage the tokens for multiple users, that's your responsibility. The ZoomClient only knows about the one token that you provide when you instantiate the client like in this example:
```csharp
var connectionInfo = new OAuthConnectionInfo(clientId, clientSecret, authorizationCode);
var client = new ZoomClient(connectionInfo);
```

In this example, the client is only aware of the one `authorizationCode`.

**Comment by ashilon** [comment]
I see. Would it be too much for ZoomClient to manage this responsibility too? Perhaps extending OAuthConnectionInfo to handle this too?

**Comment by Jericho** [comment]
That's probably not going to happen. The client would have to keep track of all the tokens and additionally the developer would have to indicate which token he wants to use whenever a method (such as `Meeting.Create`, for example) is invoked. I don't think that would be practical. But I want to keep an open mind, maybe you have an idea how you see this feature implemented?

**Comment by Jericho** [comment]
I will close this issue but feel free to add comments if you want to continue this discussion and I will be happy to reopen.
